File: src/polymomentum/data/ctf.py
"""Polymarket Conditional Token Framework (CTF) reader.

Reads on-chain market resolution state directly from Polygon, bypassing
Gamma API indexing lag. Authoritative — the CTF contract IS where market
resolution lives once UMA's optimistic oracle settles.

For a binary "Up or Down" candle market with `condition_id`:
  payoutDenominator(cid) == 0           -> not resolved yet
  payoutDenominator(cid) >  0           -> resolved
    payoutNumerators(cid, 0) > num[1]  -> outcome 0 (Up) won
    payoutNumerators(cid, 1) > num[0]  -> outcome 1 (Down) won
    50/50 split                         -> tie (rare, UMA "p1=0.5")

Pattern: raw eth_call via httpx, matching execution/wallet.py — no web3
dependency required.
"""
from __future__ import annotations

import logging

import httpx

from polymomentum.config import settings

log = logging.getLogger(__name__)

# Polymarket uses the standard Gnosis ConditionalTokens deployment on Polygon.
# Verified 2026-04-26: returns sensible payouts for our recently-resolved
# condition_ids (paper run 2026-04-25).
CTF_ADDRESS = "0x4D97DCd97eC945f40cF65F87097ACe5EA0476045"

# 4-byte selectors (keccak256 of the signature)
_PAYOUT_DENOMINATOR = "0xdd34de67"   # payoutDenominator(bytes32)
_PAYOUT_NUMERATORS = "0x0504c814"    # payoutNumerators(bytes32,uint256)


def _eth_call(data: str, rpc_url: str | None = None, timeout: float = 10.0) -> int:
    """Make a raw eth_call to the CTF contract and return the uint256 result."""
    rpc = rpc_url or settings.polygon_rpc_url
    resp = httpx.post(rpc, json={
        "jsonrpc": "2.0",
        "method": "eth_call",
        "params": [{"to": CTF_ADDRESS, "data": data}, "latest"],
        "id": 1,
    }, timeout=timeout)
    resp.raise_for_status()
    body = resp.json()
    if "error" in body:
        raise RuntimeError(f"CTF eth_call error: {body['error']}")
    return int(body.get("result", "0x0"), 16)


def _strip_0x(s: str) -> str:
    return s[2:] if s.startswith("0x") else s


def get_resolution(
    condition_id: str,
    rpc_url: str | None = None,
) -> tuple[str | None, list[int]]:
    """Read the on-chain resolution for a binary candle market.

    Returns:
        (winner, [num0, num1])
        winner is "up", "down", "tie", or None if not yet resolved.
        num0/num1 are the raw payout numerators (typically 0 or 1 for a
        binary resolution; can be fractional for UMA p1=0.5 ties).

    Assumes outcome 0 maps to "Up" (YES) and outcome 1 to "Down" (NO),
    which is how the candle scanner reads markets via Gamma. If a market
    has outcomes in the opposite order, the caller is responsible for
    reversing the result.

    Raises on RPC errors so callers can decide whether to retry.
    """
    cid_hex = _strip_0x(condition_id)
    if len(cid_hex) != 64:
        raise ValueError(f"condition_id must be 32 bytes, got {len(cid_hex)//2}")

    denom = _eth_call(_PAYOUT_DENOMINATOR + cid_hex, rpc_url)
    if denom == 0:
        return None, [0, 0]

    num0 = _eth_call(_PAYOUT_NUMERATORS + cid_hex + "0".zfill(64), rpc_url)
    num1 = _eth_call(_PAYOUT_NUMERATORS + cid_hex + "1".zfill(64), rpc_url)

    if num0 == num1:
        return "tie", [num0, num1]
    return ("up" if num0 > num1 else "down"), [num0, num1]

File: rust_engine/src/data/ctf.rs
//! Polymarket Conditional Token Framework (CTF) reader.
//!
//! Reads on-chain market resolution from Polygon.
//!
//! For binary markets:
//!   payoutDenominator(cid) == 0 → not resolved
//!   payoutNumerators(cid, 0) > num[1] → outcome 0 (Up) won
//!   num0 == num1 → tie

use anyhow::{anyhow, Context, Result};
use reqwest::Client;
use serde_json::json;

pub const CTF_ADDRESS: &str = "0x4D97DCd97eC945f40cF65F87097ACe5EA0476045";
const PAYOUT_DENOMINATOR: &str = "0xdd34de67";
const PAYOUT_NUMERATORS: &str = "0x0504c814";

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Resolution {
    Up,
    Down,
    Tie,
    NotResolved,
}

impl Resolution {
    pub fn as_str(&self) -> &'static str {
        match self {
            Resolution::Up => "up",
            Resolution::Down => "down",
            Resolution::Tie => "tie",
            Resolution::NotResolved => "pending",
        }
    }
}

pub struct CtfReader {
    rpc_url: String,
    http: Client,
}

impl CtfReader {
    pub fn new(rpc_url: impl Into<String>) -> Self {
        let http = Client::builder()
            .timeout(std::time::Duration::from_secs(10))
            .build()
            .expect("client");
        Self { rpc_url: rpc_url.into(), http }
    }

    async fn eth_call(&self, data: &str) -> Result<u128> {
        let body = json!({
            "jsonrpc": "2.0",
            "method": "eth_call",
            "params": [{"to": CTF_ADDRESS, "data": data}, "latest"],
            "id": 1,
        });
        let resp = self.http.post(&self.rpc_url).json(&body).send().await?;
        let json: serde_json::Value = resp.json().await.context("decode json-rpc")?;
        if let Some(err) = json.get("error") {
            return Err(anyhow!("CTF eth_call error: {err}"));
        }
        let result = json.get("result").and_then(|v| v.as_str()).unwrap_or("0x0");
        let trimmed = result.trim_start_matches("0x");
        let trimmed = if trimmed.is_empty() { "0" } else { trimmed };
        // The result fits in u128 for any sensible payout numerator/denominator,
        // but we parse as u128 from hex.
        u128::from_str_radix(trimmed, 16).context("parse hex")
    }

    pub async fn get_resolution(&self, condition_id: &str) -> Result<(Resolution, [u128; 2])> {
        let cid = condition_id.trim_start_matches("0x");
        if cid.len() != 64 {
            return Err(anyhow!(
                "condition_id must be 32 bytes (64 hex chars), got {}",
                cid.len() / 2
            ));
        }

        let denom_call = format!("{PAYOUT_DENOMINATOR}{cid}");
        let denom = self.eth_call(&denom_call).await?;
        if denom == 0 {
            return Ok((Resolution::NotResolved, [0, 0]));
        }

        let num0_call = format!(
            "{PAYOUT_NUMERATORS}{cid}{}",
            "0".repeat(64),
        );
        let num1_call = format!(
            "{PAYOUT_NUMERATORS}{cid}{}",
            format!("{:0>64}", "1"),
        );
        let num0 = self.eth_call(&num0_call).await?;
        let num1 = self.eth_call(&num1_call).await?;
        let res = if num0 == num1 {
            Resolution::Tie
        } else if num0 > num1 {
            Resolution::Up
        } else {
            Resolution::Down
        };
        Ok((res, [num0, num1]))
    }
}
